<!DOCTYPE html>
{% load static %}
<html lang="en">
<head>
    <meta charset="UTF-8">
        <!-- Favicon -->
    <link rel="shortcut icon" href="{% static 'favicon.ico' %}" type="image/x-icon">

    <!-- Web Fonts -->
    <link href="//fonts.googleapis.com/css?family=Playfair+Display:400,700%7COpen+Sans:300,400,700" rel="stylesheet">

    <!-- Bootstrap Styles -->
    <link rel="stylesheet" type="text/css" href="{% static 'bootstrap.min.css' %}">

    <!-- Components Vendor Styles -->
    <script src="https://kit.fontawesome.com/020635d014.js" crossorigin="anonymous"></script>

    <!-- Theme Styles -->
    <link rel="stylesheet" type="text/css" href="{% static 'styles.css' %}">

    <title>Tutorial</title>
</head>
<body data-spy="scroll" data-target="#bs-example-navbar-collapse-1" data-offset="100">
    <!-- Header -->
    <header role="header">
      <!-- Navbar -->
      <nav class="js-navbar-scroll navbar fixed-top navbar-expand-lg navbar-dark bg-primary">
        <div class="container-fluid">
          <a class="navbar-brand" href="/database/">
            HSDatabase
          </a>

          <button class="navbar-toggler" type="button"
                  data-toggle="collapse"
                  data-target="#navbarTogglerDemo"

                  aria-controls="navbarTogglerDemo"
                  aria-expanded="false"
                  aria-label="Toggle navigation">
            <span class="navbar-toggler-icon"></span>
          </button>
            <div class="collapse navbar-collapse" id="navbarTogglerDemo">

            <ul class="navbar-nav ml-auto mt-2 mt-lg-0">
              <li class="nav-item mr-4 mb-2 mb-lg-0">
                <a class="nav-link" href="/database/">Home</a>
              </li>
              <li class="nav-item mr-4 mb-2 mb-lg-0">
                <a class="nav-link" href="/database/browse/">Browse</a>
              </li>
              <li class="nav-item mr-4 mb-2 mb-lg-0">
                <a class="nav-link" href="/database/search/">Search</a>
              </li>
                <li class="nav-item mr-4 mb-2 mb-lg-0">
                <a class="nav-link" href="/database/blast/">Blast</a>
              </li>
                <li class="nav-item mr-4 mb-2 mb-lg-0">
                <a class="nav-link" href="/database/kegg/">KEGG</a>
              </li>
                <li class="nav-item mr-4 mb-2 mb-lg-0">
                <a class="nav-link active" href="/database/faq/">Tutorial</a>
              </li>
            </ul>
          </div>
        </div>
      </nav>
      <!-- End Navbar -->
    </header>
    <!-- End Header -->

    <main role="main">
    <section class="u-content-space">
        <div class="container">
            <h2>Tutorial for HSDatabase&nbsp;</h2><b><a href="/example/download/HSDatabase_Tutorials.pdf/" style="text-decoration: underline">[Download Tutorial for HSDatabase.pdf]</a></b>
            <div class="col-lg-12 align-self-center">
              <img class="img-fluid rounded" src="{% static 'HSDatabase.png' %}" alt="Iamge Description">
            <hr class="my-7">
            </div>
            <div>
                {% load embed_video_tags %}
                <p class="lead">
                    <br>
                    <strong>Webinars: Using Bioinformatics Tools to Predict, Collect and Visualize the Highly Similar Duplicates in Eukaryotic Genomes</strong>
                </p>
                {% video head_video.video 'small' %}
                <p class="lead"><br><br>
                    <strong>HSDatabase </strong>- a database of highly similar duplicate genes in eukaryotic genomes. (<a href="http://hsdfinder.com/database/">http://hsdfinder.com/database/</a>)
                </p><p>
                    &emsp;Series 4: How to use the HSDatabase - browse and search?
                    <br>&emsp;Series 5: How to use the HSDatabase - BLAST and KEGG?
                </p>
                    {% for i in videos_db %}
                        {% video i.video '350x261' %}
                    {% endfor %}
            </div>
        <hr class="my-7">
        </div>
{#    </section>#}
{#        <section class="u-content-space">#}
            <div class="container">
                {% for type, faq in faqs.items %}
                <div class="panel-group" id="accordion" role="tablist" aria-multiselectable="true">
                <p class="lead">
                    <b>FAQ</b>
                </p>
{#                    {% for item in faq %}#}
{#                  <p><div class="card">#}
{#                    <div class="card-header" role="tab" id="heading_{{ item.t }}">#}
{#                        <a role="button" data-toggle="collapse" data-parent="#accordion" href="#collapse_{{ item.t }}" aria-expanded="true" aria-controls="collapseOne">#}
{#                          {{ item.q }}#}
{#                        </a>#}
{#                    </div>#}
{#                    <div id="collapse_{{ item.t }}" class="panel-collapse collapse in" role="tabpanel" aria-labelledby="heading_{{ item.t }}">#}
{#                      <div class="card-body">#}
{#                        {{ item.a }}#}
{#                      </div>#}
{#                    </div>#}
{#                  </div></p>#}
{#                        {% endfor %}#}
                  <p><div class="card">
                    <div class="card-header" role="tab" id="heading_1">
                        <a role="button" data-toggle="collapse" data-parent="#accordion" href="#collapse_1" aria-expanded="true" aria-controls="collapseOne">
                          What’s HSDatabase?
                        </a>
                    </div>
                    <div id="collapse_1" class="panel-collapse collapse in" role="tabpanel" aria-labelledby="heading_1">
                      <div class="card-body">
                        There are many features in HSDatabase including, but not limited to the information of HSDs number, gene copy number and gene copy length. The protein function domains and pathway of HSDs are also collected from database of KEGG and InterProScan. What’s more, a BLAST search option is provided for users to conveniently explore potential homologous sequences of interest. The HSDatabase aims to become a useful platform for the prediction and comparative analysis of HSDs in the eukaryotic genomes of different survival environments, which might deepen our insights into the gene duplication mechanisms driving the genome adaptation.
                      </div>
                    </div>
                  </div></p>
                    <p><div class="card">
                    <div class="card-header" role="tab" id="heading_2">
                        <a role="button" data-toggle="collapse" data-parent="#accordion" href="#collapse_2" aria-expanded="true" aria-controls="collapseOne">
                          How you collect the HSDs in the HSDatabase?
                        </a>
                    </div>
                    <div id="collapse_2" class="panel-collapse collapse in" role="tabpanel" aria-labelledby="heading_2">
                      <div class="card-body">
                        The predicted HSDs will be manually curated before submitting to the HSDatabase. We filtered possible false positive HSDs with unfit gene copy length and add novel HSDs if necessary, especially, for those satisfy the criteria of near-identical protein lengths (within 10 amino acids and >80% pairwise identities). HSDatabase is built using a relational database (MySQL) allowing rapid retrieval of data and making resource easily maintainable. One entry corresponds to one eukaryote genome. The genomes can be browsed via the categories of plant and animal or the phylogeny for specific genome target. The genome entry is then split into many subcategories of HSDs entries. The database access is via a web interface based on python script and provides various ways to search for HSDs entries, such as organism’s name, HSDs accession number, and gene copy name, etc. HSDatabase can allow users to conveniently BLAST a query to find a homologous match. HSDatabase will be updated timely and the latest version is HSDatabase v1.5.
                      </div>
                    </div>
                  </div></p>
                    <p><div class="card">
                    <div class="card-header" role="tab" id="heading_3">
                        <a role="button" data-toggle="collapse" data-parent="#accordion" href="#collapse_3" aria-expanded="true" aria-controls="collapseOne">
                          How many species you have collected?
                        </a>
                    </div>
                    <div id="collapse_3" class="panel-collapse collapse in" role="tabpanel" aria-labelledby="heading_3">
                      <div class="card-body">
                        HSDatabase is a database of categorizing these highly similar duplicate genes (HSDs) in nuclear eukaryotes, which documented total of 42,884 HSDs in 15 eukaryotic genomes.
                      </div>
                    </div>
                  </div></p>
{#                <hr class="my-7">#}
                </div>
                {% endfor %}
            </div>
        </section>
    </main>


{#    <!-- Footer -->#}
{#    <footer class="bg-dark py-2">#}
{#      <div class="container-fluid">#}
{#          <div class="text-center">#}
{#            <small class="text-white">&copy; 2020 <a class="text-white" href="">HSDFinder</a>. All Rights Reserved.</small>#}
{#        </div>#}
{#      </div>#}
{#    </footer>#}
    <!-- JAVASCRIPTS (Load javascripts at bottom, this will reduce page load time) -->
    <!-- Global Vendor -->
    <script src="{% static 'jquery.min.js' %}"></script>
    <script src="{% static 'jquery.migrate.min.js' %}"></script>
    <script src="{% static 'popper.min.js' %}"></script>
    <script src="{% static 'bootstrap.min.js' %}"></script>

    <!-- Components Vendor  -->
    <script src="{% static 'jquery.magnific-popup.min.js' %}"></script>
    <script src="{% static 'jquery.parallax.js' %}"></script>

    <!-- Theme Settings and Calls -->
    <script src="{% static 'global.js' %}"></script>

    <!-- Theme Components and Settings -->
    <script src="{% static 'parallax.js' %}"></script>

</body>
</html>